### PR TITLE
Prelude/CBOR

### PR DESCRIPTION
**Pull request by ehmry**
Package for working with CBOR items.

---

There is no support for binary items, but I think there there is a reasonable workaround here. If a `Text` value is tagged as base64 (`CBOR/base64 : Text -> CBOR`) then a utility that converts Dhall to CBOR would be able to decode the base64 into a byte string as it encodes the CBOR binary stream. This converter could also tag the bytes string as encodable to base64, so the CBOR binary could then be transparently converted back to Dhall. See https://tools.ietf.org/html/rfc7049#section-2.4.4 for more information.

Ref #1121

### CODE REVIEW

**Comment by ehmry** [comment]
`CBOR/toJSON` is missing because I didn't figure out how to implement
`{ mapKey : CBOR, mapValue : CBOR } -> { mapKey : Text, mapValue : JSON }`.

**Comment by Nadrieril** [comment]
I think `CBOR/toJSON` would have to fail on an object key that is not a string. So you get a `{ mapKey : CBOR, mapValue : CBOR } -> Optional { mapKey : Text, mapValue : CBOR }`, and then you recursively call `CBOR/toJSON` to get `Optional { mapKey : Text, mapValue : JSON }`. But handling all the `Optional`s will probably be annoying

File: Prelude/CBOR/Type.dhall [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{-
```
Comment (Gabriel439) [inline Prelude/CBOR/Type.dhall:1]:
```suggestion
{-|
```

File: Prelude/CBOR/textMap.dhall [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    = λ(xs : List TextEntry) →
+        CBOR/map
+          ( List/map
+              TextEntry
+              { mapKey : CBOR, mapValue : CBOR }
+              (λ(x : TextEntry) → x with mapKey = CBOR/text x.mapKey)
+              xs
+          )
+
+in  mapText
```
Comment (Gabriel439) [inline Prelude/CBOR/textMap.dhall:28]:
```suggestion
let textMap
    : List TextEntry → CBOR
    = λ(xs : List TextEntry) →
        CBOR/map
          ( List/map
              TextEntry
              { mapKey : CBOR, mapValue : CBOR }
              (λ(x : TextEntry) → x with mapKey = CBOR/text x.mapKey)
              xs
          )

in  textMap
```

File: Prelude/CBOR/data.dhall [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{-|
+Create a CBOR item that incapsulates CBOR data
+
+A Dhall to CBOR encoder should convert the inner item to a byte string
+and tag it as enclosed CBOR data. This is for the benefit of a decoding
+process that is not concerned with the semantics of the inner data.
+See RFC7049 section 2.4.4.1 for more information.
+-}
+let tag = ./tag.dhall let data = tag 24 in data
```
Comment (Gabriel439) [inline Prelude/CBOR/data.dhall:9]: If you add a type signature for `data` then it should prefer to format in the multi-line form

File: Prelude/CBOR/package.dhall [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+, bool = ./bool.dhall
+, data = ./data.dhall
+, diagnostic = ./diagnostic.dhall
+, double = ./double.dhall
+, fromJSON = ./fromJSON.dhall
+, integer = ./integer.dhall
+, map = ./map.dhall
+, null = ./null.dhall
+, tag = ./tag.dhall
+, toArray = ./toArray.dhall
```
Comment (Gabriel439) [inline Prelude/CBOR/package.dhall:14]: Minor suggestion: perhaps rename `toArray` to `arrayOf`, as in `arrayOf Integer CBOR/integer`

**Review by Gabriel439 — COMMENTED** [review]
Looks great so far!

File: Prelude/CBOR/textMap.dhall (hunk 2 above)
Comment (ehmry) [reply Prelude/CBOR/textMap.dhall:28]: Yes, `textMap` was once `mapText`. `textMap` is better?

Comment (Gabriel439) [reply Prelude/CBOR/textMap.dhall]: I'm fine with either one.  I was only pointing it out because it was inconsistent with the file name

**Comment by Gabriel439** [comment]
Also, I think this is probably fine to merge in its present state.  We can always add support for converting CBOR to JSON later.

**Comment by ehmry** [comment]
On Donnerstag, 10. Dezember 2020 06:42:28 CET, Gabriel Gonzalez wrote:
> Also, I think this is probably fine to merge in its present 
> state.

I would rather wait until I get experience using it, but if there is
someone that needs it now, I'm not opposed to merging.

**Comment by Gabriel439** [comment]
@ehmry: Yeah, that's fine.  No rush

**Comment by Gabriel439** [comment]
I'll close this since it looks like this has stalled, but feel free to reopen this later if you want to revisit this

**Pull request closed without merging**